### PR TITLE
Properly synchronize threads when testing

### PR DESCRIPTION
**Pull request by Fylwind**
~~Pipes pipes pies pipes~~

Edit: actually, it's about thread synchronization.

### CODE REVIEW

**Comment by Fylwind** [comment]
(Don't merge yet, WIP)

File: src/rope.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    }
+    return s;
+}
+
+void
+clear_rope(rope *r)
+{
+    struct rope_segment *s;
+    while ((s = shift_rope(r))) {
+        free(s);
```
Comment (staticfox) [inline src/rope.c]: `xfree();`

File: src/foxbot.c [hunk 2]
```diff
@@ -130,47 +123,72 @@ parse_opts(int argc, char **argv)
...
         snprintf(buf, sizeof(buf), "Exiting due to %s", reason);
         do_quit(buf);
     }
 
-    return 0;
+    clear_channels();
+    destroy_socket();
+
+    if (bot.msg)
+        free(bot.msg->from);
```
Comment (staticfox) [inline src/foxbot.c]: `xfree();`

File: src/rope.c (hunk 1 above)
Comment (Fylwind) [inline src/rope.c]: ☑

File: src/foxbot.c (hunk 2 above)
Comment (Fylwind) [inline src/foxbot.c]: ☑

**Comment by Fylwind** [comment]
Done I think \o/

File: src/socket.c [hunk 3]
```diff
@@ -114,59 +126,94 @@ raw(char *fmt, ...)
...
-        return;
-    }
+void
+reset_io(io_state *io)
+{
+    io->_fd = -1;
+    io->_buf_used = 0;
+    io->_buf_size = 0;
+    io->_line_start = NULL;
+    free(io->_buf);
```
Comment (staticfox) [inline src/socket.c]: `xfree`

**Pull request merged**